File: src/ru/maslov/arrays/Permutations.java
package ru.maslov.arrays;

import java.util.ArrayList;
import java.util.List;

public class Permutations {
    public static void main(String[] args) {
        int[] r = {1, 2, 3};
        new Permutations().permute(r);
    }

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> permutations = new ArrayList<>();
        if (nums.length == 0) {
            return permutations;
        }

        collectPermutations(nums, new ArrayList<>(), permutations);
        return permutations;
    }

    private void collectPermutations(int[] nums, List<Integer> permutation, List<List<Integer>> permutations) {
        int s = permutation.size();

        if (s == nums.length) {
            permutations.add(permutation);
            return;
        }

        int el = nums[s];

        for (int position = 0; position <= s; position++) {
            List<Integer> newPermutation = new ArrayList<>(permutation);
            newPermutation.add(position, el);
            collectPermutations(nums, newPermutation, permutations);
        }
    }
}

File: src/ru/maslov/kotlin/Permutations.kt
package ru.maslov.kotlin

import java.util.*

fun main(args: Array<String>) {
    val x: IntArray = intArrayOf(1, 2, 3)

    permute(x)
}

fun permute(nums: IntArray): List<List<Int>> {
    val permutations = ArrayList<ArrayList<Int>>()
    if (nums.size == 0) {
        return permutations
    }

    collectPermutations(nums, ArrayList<Int>(), permutations)
    return permutations
}

fun collectPermutations(nums: IntArray, permutation: ArrayList<Int>, permutations: ArrayList<ArrayList<Int>>) {
    val ps = permutation.size

    if (ps == nums.size) {
        permutations.add(permutation)
        return
    }
    val el = nums.get(ps)

    for (position in 0..ps) {
        val _permutation = ArrayList<Int>(permutation)
        _permutation.add(position, el)
        collectPermutations(nums, _permutation, permutations)
    }
}